staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go

```diff
@@ -383,51 +383,98 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
 		if pool.Generation > generation {
 			generation = pool.Generation
 		}
-
-		// Right now all devices get published in a single slice.
-		// We simply pick the first one, if there is one, and copy
-		// it in preparation for updating it.
-		//
-		// TODO: support splitting across slices, with unit tests.
-		if len(slices) > 0 {
-			obsoleteSlices = append(obsoleteSlices, slices[1:]...)
-			slices = []*resourceapi.ResourceSlice{slices[0].DeepCopy()}
-		} else {
-			slices = []*resourceapi.ResourceSlice{
-				{
+		needToAdd, needToDelete := getDevicesToAddAndDelete(pool, slices)
+		c.mutex.Lock()
+		slices = removeDevicesFromSlice(slices, needToDelete)
```

Comment:
This looks suspicious. `slices` is a local variable in this method. What needs to be protected by the mutex here?